### PR TITLE
Add title, frontend_title to membership type

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
I hit a situation where I wanted the membership type name but thought it should be localised & wasn't sure what membership.membership_type_id.name would do

Then I remembered @mattwire wanted to add label anyway so I thought I'd take a stab. I went with title cos I figured we should also add frontend_title & that maybe made the case for title over label.

However - I found I can no longer run regen.sh & the error is on a line that has a today-commit by @totten 
```
found 0 vulnerabilities
+ popd
/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/bin
+ '[' -n 1 -a -d /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/bin/./../xml ']'
+ pushd /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/bin/./../xml
/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/xml /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/bin
+ '[' -z 3306 ']'
+ PHP_MYSQL_HOSTPORT=database:3306
+ php -d mysql.default_host=database:3306 -d mysql.default_user=dmastercivicrm -d mysql.default_password=dmastercivicrm GenCode.php schema/Schema.xml '' Drupal

civicrm_domain.version := 5.69.alpha1

Parsing schema description schema/Schema.xml
Extracting database information
Extracting table information
Generating civicrm.config.php
Generating table list
Generating sql file
Generating sql drop tables file
Generating data files for en_US
Generating navigation file
Generating sample file
PHP Warning:  Smarty error: unable to read resource: &quot;languages_setup.tpl&quot; in /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/packages/Smarty/Smarty.class.php on line 1100
PHP Stack trace:
PHP   1. {main}() /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/xml/GenCode.php:0
PHP   2. CRM_Core_CodeGen_Main->main() /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/xml/GenCode.php:58
PHP   3. CRM_Core_CodeGen_I18n->run() /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/CodeGen/Main.php:132
PHP   4. CRM_Core_CodeGen_I18n->generateInstallLangs() /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/CodeGen/I18n.php:9
PHP   5. CRM_Core_CodeGen_Util_Template->run($infile = 'languages_setup.tpl', $outpath = '../setup/res/languages.php') /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/CodeGen/I18n.php:28
PHP   6. Smarty->fetch($resource_name = 'languages_setup.tpl', $cache_id = *uninitialized*, $compile_id = *uninitialized*, $display = *uninitialized*) /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/CodeGen/Util/Template.php:67
PHP   7. Smarty->_compile_resource($resource_name = 'languages_setup.tpl', $compile_path = '/tmp/templates_c_N5SMun.d/%%F4^F41^F41B56F0%%languages_setup.tpl.php') /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/packages/Smarty/Smarty.class.php:1289
PHP   8. Smarty->_fetch_resource_info($params = ['resource_name' => 'languages_setup.tpl', 'get_source' => TRUE, 'quiet' => FALSE]) /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/packages/Smarty/Smarty.class.php:1464
PHP   9. Smarty->trigger_error($error_msg = 'unable to read resource: "languages_setup.tpl"', $error_type = *uninitialized*) /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/packages/Smarty/Smarty.class.php:1673
PHP  10. trigger_error($message = 'Smarty error: unable to read resource: &quot;languages_setup.tpl&quot;', $error_type = 512) /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/packages/Smarty/Smarty.class.php:1100
PHP Notice:  Undefined property: PHP_Beautifier::$sFile in /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/packages/PHP/Beautifier.php on line 758
PHP Stack trace:
PHP   1. {main}() /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/xml/GenCode.php:0
PHP   2. CRM_Core_CodeGen_Main->main() /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/xml/GenCode.php:58
PHP   3. CRM_Core_CodeGen_I18n->run() /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/CodeGen/Main.php:132
PHP   4. CRM_Core_CodeGen_I18n->generateInstallLangs() /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/CodeGen/I18n.php:9
PHP   5. CRM_Core_CodeGen_Util_Template->run($infile = 'languages_setup.tpl', $outpath = '../setup/res/languages.php') /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/CodeGen/I18n.php:28
PHP   6. PHP_Beautifier->process() /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/CodeGen/Util/Template.php:71
PHP Fatal error:  Uncaught Exception: Nothing on output! in /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/packages/PHP/Beautifier.php:761
Stack trace:
#0 /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/CodeGen/Util/Template.php(71): PHP_Beautifier->process()
#1 /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/CodeGen/I18n.php(28): CRM_Core_CodeGen_Util_Template->run()
#2 /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/CodeGen/I18n.php(9): CRM_Core_CodeGen_I18n->generateInstallLangs()
#3 /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/CodeGen/Main.php(132): CRM_Core_CodeGen_I18n->run()
#4 /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/xml/GenCode.php(58): CRM_Core_CodeGen_Main->main()
#5 {main}
  thrown in /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/packages/PHP/Beautifier.php on line 761
```
Overview
----------------------------------------
_A brief description of the pull request. Keep technical jargon to a minimum. Hyperlink relevant discussions._

Before
----------------------------------------
_What is the old user-interface or technical-contract (as appropriate)?_
_For optimal clarity, include a concrete example such as a screenshot, GIF ([LICEcap](http://www.cockos.com/licecap/), [SilentCast](https://github.com/colinkeenan/silentcast)), or code-snippet._

After
----------------------------------------
_What changed? What is new old user-interface or technical-contract?_
_For optimal clarity, include a concrete example such as a screenshot, GIF ([LICEcap](http://www.cockos.com/licecap/), [SilentCast](https://github.com/colinkeenan/silentcast)), or code-snippet._

Technical Details
----------------------------------------
_If the PR involves technical details/changes/considerations which would not be manifest to a casual developer skimming the above sections, please describe the details here._

Comments
----------------------------------------
_Anything else you would like the reviewer to note_

### CODE REVIEW

**Comment by mattwire** [comment]
@eileenmcnaughton Most of the tests are failing because title and frontend_title are required fields. I tried and failed to set them via the pre hook in the membershipType BAO because DAOSaveAction calls `validateValues()` which runs before the hook and fails.
I can't understand what is the difference with MembershipType compared with some other entities where frontend_title and title were added.

File: CRM/Upgrade/Incremental/php/FiveSixtyNine.php [hunk 1]
```diff
@@ -37,6 +43,16 @@ public function upgrade_5_69_alpha1($rev): void {
...
+    $this->addTask('Make civicrm_membership.title required', 'alterColumn', 'civicrm_membership_type', 'title',
+      "varchar(255) COLLATE utf8mb4_unicode_ci NULL COMMENT 'Membership Type title. For top of page display"
+    );
+    $this->addTask('Add MembershipType.frontend_title', 'alterColumn', 'civicrm_membership_type', 'frontend_title',
+      "varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL COMMENT 'Membership Type title'"
+    );
+    $this->addTask('Add MembershipType.frontend_title', 'addColumn', 'civicrm_membership_type', 'frontend_title',
+      "varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL COMMENT 'Membership Type Public title'"
+    );
+    $this->addTask(ts('Create index %1', [1 => 'civicrm_membership_type.UI_name']), 'addIndex', 'civicrm_membership_type', [['name']], 'UI');
```
Comment (mattwire) [inline CRM/Upgrade/Incremental/php/FiveSixtyNine.php:55]: @eileenmcnaughton I think somethings gone a bit wrong here. You've got duplicate upgrade tasks.

Comment (eileenmcnaughton) [reply CRM/Upgrade/Incremental/php/FiveSixtyNine.php:55]: @mattwire one task to add the fields then after the sql runs one to make them required

**Comment by eileenmcnaughton** [comment]
api_v3_ContributionPageTest::testSubmitMembershipBlockNotSeparatePaymentProcessorInstantRenew
Failure in api call for MembershipType Create:  DB Error: unknown error
 INSERT INTO `civicrm_membership_type` (`domain_id` , `name` , `member_of_contact_id` , `financial_type_id` , `minimum_fee` , `duration_unit` , `duration_interval` , `period_type` , `visibility` , `is_active` ) VALUES ( 1 , 'General' ,  4 ,  2 ,  2 , 'year' ,  1 , 'rolling' , 'Public' ,  1 )

**Comment by eileenmcnaughton** [comment]
it's picking up frontend_title but not title now

api_v3_ContactTest::testMergeOrganizations
Failure in api call for membership_type create:  DB Error: unknown error
 INSERT INTO `civicrm_membership_type` (`domain_id` , `frontend_title` , `name` , `member_of_contact_id` , `financial_type_id` , `minimum_fee` , `duration_unit` , `duration_interval` , `period_type` , `relationship_type_id` , `relationship_direction` , `visibility` , `is_active` ) VALUES ( 1 , 'Organization Membership' , 'Organization Membership' ,  1 ,  1 ,  10 , 'year' ,  1 , 'rolling' , '5' , 'b_a' , 'Public' ,  1 )

**Comment by mattwire** [comment]
@eileenmcnaughton I've rebased and pushed this. There was some duplication and a missing \` so I've fixed that. However, I'm not sure how it is meant to work between the php upgrade steps and the tpl(copy name to title/frontend_title) steps because we need to fill title/frontend_title before making the fields required.

**Comment by mattwire** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
There are still quite a few challenges on this - I'm gonna close it out for now - this PR https://github.com/civicrm/civicrm-core/pull/28680 might fix some of the test issues (which are mostly about the tests setting up 2 membership types with the same name)

File: CRM/Upgrade/Incremental/php/FiveSeventy.php [hunk 2]
```diff
@@ -27,8 +27,24 @@ class CRM_Upgrade_Incremental_php_FiveSeventy extends CRM_Upgrade_Incremental_Ba
    * @param string $rev
    *   The version number matching this function name
    */
-  public function upgrade_5_70_alpha1($rev): void {
+  public function upgrade_5_70_alpha1(string $rev): void {
+    $this->addTask('Add MembershipType.title', 'addColumn', 'civicrm_membership_type', 'title',
+      "varchar(255) COLLATE utf8mb4_unicode_ci NULL COMMENT 'Membership Type title. For top of page display"
```
Comment (eileenmcnaughton) [inline CRM/Upgrade/Incremental/php/FiveSeventy.php:32]: missing single quote at end

**Pull request closed without merging**